File: Task/Kirkpatrick-Seidel-algorithm/Elixir/kirkpatrick-seidel-algorithm.ex
defmodule ConvexHullAlgorithm do
  defmodule Point do
    @enforce_keys [:x, :y]
    defstruct [:x, :y]

    def new(x \\ 0, y \\ 0) do
      %Point{x: x, y: y}
    end

    def compare(%Point{x: x1, y: y1}, %Point{x: x2, y: y2}) do
      cond do
        x1 == x2 -> compare_float(y1, y2)
        true -> compare_float(x1, x2)
      end
    end

    def compare_float(a, b) do
      cond do
        a < b -> -1
        a > b -> 1
        true -> 0
      end
    end

    def less_than(%Point{x: x1, y: y1}, %Point{x: x2, y: y2}) do
      cond do
        x1 == x2 -> y1 < y2
        true -> x1 < x2
      end
    end

    def greater_than(%Point{x: x1, y: y1}, %Point{x: x2, y: y2}) do
      cond do
        x1 == x2 -> y1 > y2
        true -> x1 > x2
      end
    end
  end

  def flipped(points) do
    Enum.map(points, fn point ->
      Point.new(-point.x, -point.y)
    end)
  end

  def quick_select(list, index) do
    quick_select(list, index, 0, length(list) - 1)
  end

  def quick_select(list, _index, lo, hi) when lo == hi do
    Enum.at(list, lo)
  end

  def quick_select(list, index, lo, hi) do
    pivot_index = lo + :rand.uniform(hi - lo + 1) - 1
    pivot_value = Enum.at(list, pivot_index)
    list = swap(list, lo, pivot_index)

    {list, cur} = Enum.reduce(lo + 1..hi, {list, lo}, fn run, {acc_list, acc_cur} ->
      if compare_elements(Enum.at(acc_list, run), pivot_value) < 0 do
        {swap(acc_list, acc_cur + 1, run), acc_cur + 1}
      else
        {acc_list, acc_cur}
      end
    end)

    list = swap(list, cur, lo)

    cond do
      index < cur -> quick_select(list, index, lo, cur - 1)
      index > cur -> quick_select(list, index, cur + 1, hi)
      true -> Enum.at(list, cur)
    end
  end

  defp compare_elements(a, b) when is_float(a) and is_float(b), do: Point.compare_float(a, b)
  defp compare_elements(%Point{} = a, %Point{} = b), do: Point.compare(a, b)

  defp swap(list, i, j) do
    temp = Enum.at(list, i)
    list = List.replace_at(list, i, Enum.at(list, j))
    List.replace_at(list, j, temp)
  end

  def bridge(points, vertical_line) do
    points_list = MapSet.to_list(points)

    if MapSet.size(points) == 2 do
      sorted_points = Enum.sort(points_list, &(Point.compare(&1, &2) < 0))
      [Enum.at(sorted_points, 0), Enum.at(sorted_points, 1)]
    else
      candidates = MapSet.new()
      pairs = []

      {pairs, candidates} =
        points_list
        |> Enum.chunk_every(2, 2, :discard)
        |> Enum.reduce({pairs, candidates}, fn [p1, p2], {pairs_acc, candidates_acc} ->
          if Point.less_than(p1, p2) do
            {pairs_acc ++ [[p1, p2]], candidates_acc}
          else
            {pairs_acc ++ [[p2, p1]], candidates_acc}
          end
        end)

      candidates =
        if rem(length(points_list), 2) == 1 do
          MapSet.put(candidates, List.last(points_list))
        else
          candidates
        end

      {slopes, valid_pairs} =
        Enum.reduce(pairs, {[], []}, fn [p1, p2] = pair, {slopes_acc, valid_pairs_acc} ->
          if p1.x == p2.x do
            if p1.y > p2.y do
              {slopes_acc, valid_pairs_acc}
            else
              {slopes_acc, valid_pairs_acc}
            end
          else
            slope = (p1.y - p2.y) / (p1.x - p2.x)
            {slopes_acc ++ [slope], valid_pairs_acc ++ [pair]}
          end
        end)

      if Enum.empty?(slopes) do
        if MapSet.size(candidates) >= 2 do
          candidate_list = MapSet.to_list(candidates) |> Enum.sort(&(Point.compare(&1, &2) < 0))
          [List.first(candidate_list), List.last(candidate_list)]
        else
          [List.first(points_list), Enum.at(points_list, 1)]
        end
      else
        median_index = div(length(slopes), 2) - (if rem(length(slopes), 2) == 0, do: 1, else: 0)
        median_slope = quick_select(slopes, median_index)

        {small, equal, large} =
          Enum.zip(slopes, valid_pairs)
          |> Enum.reduce({[], [], []}, fn {slope, pair}, {small_acc, equal_acc, large_acc} ->
            cond do
              slope < median_slope -> {small_acc ++ [pair], equal_acc, large_acc}
              slope == median_slope -> {small_acc, equal_acc ++ [pair], large_acc}
              true -> {small_acc, equal_acc, large_acc ++ [pair]}
            end
          end)

        max_intercept =
          points_list
          |> Enum.map(fn point -> point.y - median_slope * point.x end)
          |> Enum.max()

        max_set =
          points_list
          |> Enum.filter(fn point -> point.y - median_slope * point.x == max_intercept end)

        left = Enum.min_by(max_set, &(&1.x))
        right = Enum.max_by(max_set, &(&1.x))

        if left.x <= vertical_line && right.x > vertical_line do
          [left, right]
        else
          new_candidates =
            if right.x <= vertical_line do
              candidates =
                large
                |> Enum.reduce(candidates, fn [_, p2], acc -> MapSet.put(acc, p2) end)

              candidates =
                equal
                |> Enum.reduce(candidates, fn [_, p2], acc -> MapSet.put(acc, p2) end)

              small
              |> Enum.reduce(candidates, fn [p1, p2], acc ->
                acc |> MapSet.put(p1) |> MapSet.put(p2)
              end)
            else
              candidates =
                small
                |> Enum.reduce(candidates, fn [p1, _], acc -> MapSet.put(acc, p1) end)

              candidates =
                equal
                |> Enum.reduce(candidates, fn [p1, _], acc -> MapSet.put(acc, p1) end)

              large
              |> Enum.reduce(candidates, fn [p1, p2], acc ->
                acc |> MapSet.put(p1) |> MapSet.put(p2)
              end)
            end

          bridge(new_candidates, vertical_line)
        end
      end
    end
  end

  def connect(lower, upper, points) do
    if lower == upper do
      [lower]
    else
      points_vec = MapSet.to_list(points) |> Enum.sort(&(Point.compare(&1, &2) < 0))
      mid_index = div(length(points_vec) - 1, 2)

      max_left = quick_select(points_vec, mid_index)
      min_right = quick_select(points_vec, mid_index + 1)

      [left, right] = bridge(points, (max_left.x + min_right.x) / 2)

      points_left = MapSet.new([left])
      points_right = MapSet.new([right])

      {points_left, points_right} =
        points
        |> Enum.reduce({points_left, points_right}, fn point, {left_acc, right_acc} ->
          cond do
            point.x < left.x -> {MapSet.put(left_acc, point), right_acc}
            point.x > right.x -> {left_acc, MapSet.put(right_acc, point)}
            true -> {left_acc, right_acc}
          end
        end)

      left_result = connect(lower, left, points_left)
      right_result = connect(right, upper, points_right)

      left_result ++ right_result
    end
  end

  def upper_hull(points) do
    lower = Enum.min_by(points, &(&1.x))

    # Find the lowest point with same x-coordinate as minimum
    lower =
      points
      |> Enum.filter(fn point -> point.x == lower.x && point.y > lower.y end)
      |> Enum.reduce(lower, fn point, acc ->
        if point.y > acc.y, do: point, else: acc
      end)

    upper = Enum.max_by(points, &(&1.x))

    filtered_points =
      points
      |> Enum.filter(fn point ->
        lower.x < point.x && point.x < upper.x || point == lower || point == upper
      end)
      |> MapSet.new()

    connect(lower, upper, filtered_points)
  end

  def convex_hull(points) do
    upper = upper_hull(points)

    flipped_points =
      points
      |> Enum.map(fn p -> Point.new(-p.x, -p.y) end)
      |> MapSet.new()

    flipped_upper = upper_hull(flipped_points)
    lower = flipped(flipped_upper)

    # Remove duplicate points at the start/end
    {upper, lower} =
      if Enum.count(upper) > 0 && Enum.count(lower) > 0 do
        upper_clean =
          if List.last(upper) == List.first(lower) do
            List.delete_at(upper, -1)
          else
            upper
          end

        lower_clean =
          if List.first(upper_clean) == List.last(lower) do
            List.delete_at(lower, -1)
          else
            lower
          end

        {upper_clean, lower_clean}
      else
        {upper, lower}
      end

    upper ++ lower
  end

  def main do
    points = MapSet.new([
      Point.new(0.0, 0.0),   # projection of [0.0, 0.0, 0.0]
      Point.new(1.0, 0.0),   # projection of [1.0, 0.0, 0.0]
      Point.new(0.0, 1.0),   # projection of [0.0, 1.0, 0.0]
      Point.new(0.5, 0.5)    # projection of [0.0, 0.0, 1.0] (projected to 2D)
    ])

    IO.puts("Input points:")
    points
    |> Enum.each(fn p ->
      IO.puts("(#{p.x}, #{p.y})")
    end)

    hull = convex_hull(points)

    IO.puts("\nConvex hull points:")
    hull
    |> Enum.each(fn p ->
      IO.puts("(#{p.x}, #{p.y})")
    end)
  end
end

# To run the algorithm
ConvexHullAlgorithm.main()

File: Task/Kirkpatrick-Seidel-algorithm/C-sharp/kirkpatrick-seidel-algorithm.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Diagnostics;

public class ConvexHullAlgorithm
{
    public class Point : IComparable<Point>
    {
        public double X { get; set; }
        public double Y { get; set; }

        public Point() : this(0, 0) { }

        public Point(double x, double y)
        {
            X = x;
            Y = y;
        }

        public int CompareTo(Point other)
        {
            if (X == other.X)
            {
                return Y.CompareTo(other.Y);
            }
            return X.CompareTo(other.X);
        }

        public override bool Equals(object obj)
        {
            if (obj == null || GetType() != obj.GetType())
            {
                return false;
            }
            Point other = (Point)obj;
            return X == other.X && Y == other.Y;
        }

        public override int GetHashCode()
        {
            unchecked
            {
                int hash = 17;
                hash = hash * 23 + X.GetHashCode();
                hash = hash * 23 + Y.GetHashCode();
                return hash;
            }
        }

        public static bool operator <(Point p1, Point p2)
        {
            if (p1.X == p2.X)
            {
                return p1.Y < p2.Y;
            }
            return p1.X < p2.X;
        }

        public static bool operator >(Point p1, Point p2)
        {
            if (p1.X == p2.X)
            {
                return p1.Y > p2.Y;
            }
            return p1.X > p2.X;
        }

        public static bool operator ==(Point p1, Point p2)
        {
            if (ReferenceEquals(p1, p2)) return true;
            if (ReferenceEquals(p1, null) || ReferenceEquals(p2, null)) return false;
            return p1.X == p2.X && p1.Y == p2.Y;
        }

        public static bool operator !=(Point p1, Point p2)
        {
            return !(p1 == p2);
        }
    }

    private static Random rand = new Random();

    private static List<Point> Flipped(IEnumerable<Point> points)
    {
        List<Point> result = new List<Point>();
        foreach (var point in points)
        {
            result.Add(new Point(-point.X, -point.Y));
        }
        return result;
    }

    private static T QuickSelect<T>(List<T> ls, int index, int lo = 0, int hi = -1) where T : IComparable<T>
    {
        if (hi == -1)
        {
            hi = ls.Count - 1;
        }

        if (lo == hi)
        {
            return ls[lo];
        }

        int pivotIndex = lo + rand.Next(hi - lo + 1);
        T pivotValue = ls[pivotIndex];
        Swap(ls, lo, pivotIndex);

        int cur = lo;
        for (int run = lo + 1; run <= hi; run++)
        {
            if (ls[run].CompareTo(pivotValue) < 0)
            {
                cur++;
                Swap(ls, cur, run);
            }
        }

        Swap(ls, cur, lo);

        if (index < cur)
        {
            return QuickSelect(ls, index, lo, cur - 1);
        }
        else if (index > cur)
        {
            return QuickSelect(ls, index, cur + 1, hi);
        }
        else
        {
            return ls[cur];
        }
    }

    private static void Swap<T>(List<T> list, int i, int j)
    {
        T temp = list[i];
        list[i] = list[j];
        list[j] = temp;
    }

    private static Tuple<Point, Point> Bridge(HashSet<Point> points, double verticalLine)
    {
        HashSet<Point> candidates = new HashSet<Point>();

        if (points.Count == 2)
        {
            return Tuple.Create(points.Min(), points.Max());
        }

        List<Tuple<Point, Point>> pairs = new List<Tuple<Point, Point>>();
        List<Point> modifyList = points.ToList();

        for (int i = 0; i < modifyList.Count / 2 * 2; i += 2)
        {
            Point p1 = modifyList[i];
            Point p2 = modifyList[i + 1];
            if (p1 < p2)
            {
                pairs.Add(Tuple.Create(p1, p2));
            }
            else
            {
                pairs.Add(Tuple.Create(p2, p1));
            }
        }

        if (modifyList.Count % 2 == 1)
        {
            candidates.Add(modifyList.Last());
        }

        List<double> slopes = new List<double>();
        List<Tuple<Point, Point>> validPairs = new List<Tuple<Point, Point>>();

        foreach (var pair in pairs)
        {
            if (pair.Item1.X == pair.Item2.X)
            {
                candidates.Add(pair.Item1.Y > pair.Item2.Y ? pair.Item1 : pair.Item2);
            }
            else
            {
                slopes.Add((pair.Item1.Y - pair.Item2.Y) / (pair.Item1.X - pair.Item2.X));
                validPairs.Add(pair);
            }
        }

        if (slopes.Count == 0)
        {
            if (candidates.Count >= 2)
            {
                return Tuple.Create(candidates.Min(), candidates.Max());
            }
             // If we don't have enough candidates, return the first pair
             var pointList = points.ToList();
             return Tuple.Create(pointList[0], pointList[1]);
        }


        int medianIndex = slopes.Count / 2 - (slopes.Count % 2 == 0 ? 1 : 0);
        double medianSlope = QuickSelect(slopes, medianIndex);

        HashSet<Tuple<Point, Point>> small = new HashSet<Tuple<Point, Point>>();
        HashSet<Tuple<Point, Point>> equal = new HashSet<Tuple<Point, Point>>();
        HashSet<Tuple<Point, Point>> large = new HashSet<Tuple<Point, Point>>();

        for (int i = 0; i < slopes.Count; i++)
        {
            if (slopes[i] < medianSlope)
            {
                small.Add(validPairs[i]);
            }
            else if (slopes[i] == medianSlope)
            {
                equal.Add(validPairs[i]);
            }
            else
            {
                large.Add(validPairs[i]);
            }
        }

        double maxSlope = double.NegativeInfinity;
        foreach (var point in points)
        {
            maxSlope = Math.Max(maxSlope, point.Y - medianSlope * point.X);
        }

        List<Point> maxSet = new List<Point>();
        foreach (var point in points)
        {
            if (point.Y - medianSlope * point.X == maxSlope)
            {
                maxSet.Add(point);
            }
        }

        Point left = maxSet.Min();
        Point right = maxSet.Max();

        if (left.X <= verticalLine && right.X > verticalLine)
        {
            return Tuple.Create(left, right);
        }

        if (right.X <= verticalLine)
        {
            foreach (var pair in large)
            {
                candidates.Add(pair.Item2);
            }
            foreach (var pair in equal)
            {
                candidates.Add(pair.Item2);
            }
            foreach (var pair in small)
            {
                candidates.Add(pair.Item1);
                candidates.Add(pair.Item2);
            }
        }

        if (left.X > verticalLine)
        {
            foreach (var pair in small)
            {
                candidates.Add(pair.Item1);
            }
            foreach (var pair in equal)
            {
                candidates.Add(pair.Item1);
            }
            foreach (var pair in large)
            {
                candidates.Add(pair.Item1);
                candidates.Add(pair.Item2);
            }
        }

        return Bridge(candidates, verticalLine);
    }

    private static List<Point> Connect(Point lower, Point upper, HashSet<Point> points)
    {
        if (lower == upper)
        {
            return new List<Point> { lower };
        }

        List<Point> pointsVec = points.ToList();
        pointsVec.Sort(); // Sort required for quickselect index logic

        int midIndex = (pointsVec.Count - 1) / 2;

        Point maxLeft = QuickSelect(pointsVec, midIndex);
        Point minRight = QuickSelect(pointsVec, midIndex + 1);

        var bridgePoints = Bridge(points, (maxLeft.X + minRight.X) / 2);
        Point left = bridgePoints.Item1;
        Point right = bridgePoints.Item2;

        HashSet<Point> pointsLeft = new HashSet<Point> { left };
        HashSet<Point> pointsRight = new HashSet<Point> { right };

        foreach (var point in points)
        {
            if (point.X < left.X)
            {
                pointsLeft.Add(point);
            }
            else if (point.X > right.X)
            {
                pointsRight.Add(point);
            }
        }

        List<Point> leftResult = Connect(lower, left, pointsLeft);
        List<Point> rightResult = Connect(right, upper, pointsRight);

        leftResult.AddRange(rightResult);
        return leftResult;
    }

    private static List<Point> UpperHull(HashSet<Point> points)
    {
        Point lower = points.Min();

        // Find the lowest point with the same x-coordinate as the minimum
        foreach (var point in points)
        {
            if (point.X == lower.X && point.Y > lower.Y)
            {
                lower = point;
            }
        }

        Point upper = points.Max();

        HashSet<Point> filteredPoints = new HashSet<Point> { lower, upper };
        foreach (var p in points)
        {
            if (lower.X < p.X && p.X < upper.X)
            {
                filteredPoints.Add(p);
            }
        }

        return Connect(lower, upper, filteredPoints);
    }

    public static List<Point> ConvexHull(HashSet<Point> points)
    {
        List<Point> upper = UpperHull(points);

        HashSet<Point> flippedPoints = new HashSet<Point>();
        foreach (var p in points)
        {
            flippedPoints.Add(new Point(-p.X, -p.Y));
        }

        List<Point> flippedUpper = UpperHull(flippedPoints);
        List<Point> lower = Flipped(flippedUpper);

        // Remove duplicate points at the start/end
        if (upper.Count > 0 && lower.Count > 0)
        {
            if (upper.Last().Equals(lower.First()))
            {
                upper.RemoveAt(upper.Count - 1);
            }

            if (upper.First().Equals(lower.Last()))
            {
                lower.RemoveAt(lower.Count - 1);
            }
        }


        List<Point> result = upper;
        result.AddRange(lower);

        return result;
    }

    // Test case for a simplex
    public static void Main(string[] args)
    {
        // Create points for a 2D projection of a 3D simplex
        HashSet<Point> points = new HashSet<Point>
        {
            new Point(0.0, 0.0),   // projection of [0.0, 0.0, 0.0]
            new Point(1.0, 0.0),   // projection of [1.0, 0.0, 0.0]
            new Point(0.0, 1.0),   // projection of [0.0, 1.0, 0.0]
            new Point(0.5, 0.5)    // projection of [0.0, 0.0, 1.0] (projected to 2D)
        };

        Console.WriteLine("Input points:");
        foreach (var p in points)
        {
            Console.WriteLine($"({p.X}, {p.Y})");
        }

        List<Point> hull = ConvexHull(points);

        Console.WriteLine("\nConvex hull points:");
        foreach (var p in hull)
        {
            Console.WriteLine($"({p.X}, {p.Y})");
        }
    }
}
